### DISCUSSION TITLE
Make slider bodies skinnable

### DISCUSSION THREAD

**Discussion by dthusian** [ppy/osu#27795] [Ideas]
Currently, slider bodies are the only osu! gameplay element that isn't easily skinnable. Skins can only set `SliderTrackOverride` and `SliderBorder` to control parts of the color of the slider.

Instead of that, I propose that the colors on a slider body be determined by two textures: `sliderbody.png` and `sliderbodyoverlay.png`. Both textures would have dimensions 1 x N, and would represent the colors from the center to the edge of the slider. Like hitcircles, `sliderbody.png` would be tinted with the combo color (or `SliderTrackOverride`, if present), and `sliderbodyoverlay.png` would overlaid on top as-is.

The advantages of this include skins being able to have fully transparent slider bodies or to be able to emulate argon's flat slider bodies.

I'm willing to work on this feature myself and have opened this discussion to get feedback first.

**Comment by Rekunan** [comment]
The general idea is good, but I'm having trouble understanding the details, could you provide some example images/elements?

**Reply by dthusian** [reply]
I created a diagram:

![drawing](https://github.com/ppy/osu/assets/37712025/4b9f9cc8-4a04-4547-b77c-dce4c5d1f723)

In the actual proposal, both textures are only 1 pixel tall, but they are larger here for visibility.

**Reply by Rekunan** [reply]
Thank you for that, although now I'm having trouble understanding how `sliderbodyoverlay.png` would work with `SliderBorder`. Would they interact in some way, or just have png replace the value?

**Reply by dthusian** [reply]
If any of `sliderbodyoverlay.png` or `sliderbody.png` is present, then `SliderBorder` should be ignored. The idea is that skins make their own slider borders using the two textures.

**Reply by Rekunan** [reply]
That sounds more like a jank solution, I think it'd be better if they could interact in a more seamless way, or maybe just remove `SliderBorder` for the skin.ini version this gets implemented.

**Reply by dthusian** [reply]
I wanted `sliderbody.png` and `sliderbodyoverlay.png` to mirror the system used by `hitcircle.png` and `hitcircleoverlay.png`, where there is no tint applied to `hitcircleoverlay.png`. `SliderBorder` could be used to tint `sliderbodyoverlay.png`, but there would be no difference in using that value and the skin author tinting the texture themselves. Thus, I think `SliderBorder` should be removed.

**Comment by JamieWright99** [comment]
Just thought to say McOsu lets you do this and it's a cool feature (slidergradient.png if anyone wants to try), I don't think there is a separate body and border asset to use there though. Definitely would be cool for something like this on lazer, I'm tired of using the (ugly imo) stable slider. Here's some random ones I made just to show
![slidersr](https://github.com/ppy/osu/assets/102170503/6db06c81-9fec-48f0-ab2f-cfe8b4531c34)

**Comment by cihe13375** [comment]
I support this. Today I tried some skinning stuff for the first time and I'm surprised by the fact that sliderbodies are not skinnable. Now I finally understand why so many skins have transparent circles but heavy dark sliderbodies which don't look great imo.

**Comment by minetoblend** [comment]
I messed around with this idea for fun, though instead of just a gradient I propose you should be able to use 2d textures that get drawn along the path. I made a small prototype for a variant of osu-framework's `SmoothPath` class which does just that. Skinning side is fairly simple, just provide a texture like in the left column on the table below.

Please don't mind the textures I'm using here, it's just whatever textures I happened to have downloaded that were vertically tileable

|source image|slider|
|-|-|
|<img width="256" height="512" alt="candy-cane" src="https://github.com/user-attachments/assets/a95e459e-c050-4921-bbcd-d9ac382dccc7" />|<img width="1131" height="818" alt="image" src="https://github.com/user-attachments/assets/9bbd4c6a-b3c3-448d-97b6-6c7850565bcd" />|
|<img width="512" height="1024" alt="bark" src="https://github.com/user-attachments/assets/820c2a63-fb66-40f4-b5b7-2d1d72321e8c" />|<img width="2555" height="1339" alt="image" src="https://github.com/user-attachments/assets/f3688b4c-bdf9-4d03-b388-4c0b2bcf7450" />|

Fun fact: this can also be used to run doom on a slider

https://github.com/user-attachments/assets/f38423ca-131d-4db1-ace7-1cee005665a8

**Reply by Natelytle** [reply]
Example texture I made that resembles argon using this (courtesy of marvin since I don't want to set this up), looks great to me!

<img width="974" height="830" alt="image" src="https://github.com/user-attachments/assets/ce154c04-518e-4a15-b084-e1a88fb6c329" />

<img width="128" height="256" alt="image" src="https://github.com/user-attachments/assets/62a02e31-d8d0-4328-93e4-e6f33be9508d" />

**Comment by I-AmBen** [comment]
Bump

**Comment by R3Ked** [comment]
I like this idea, honestly the current slider bodies look kind of ugly in my opinion and I'd like to be able to change them. If nothing else, at least give us the ability to switch a skin to rendering sliders how Argon does it.
